remove unfree HDCP blob from arm-trusted-firmware, closes #148890

a-m-joseph opened the pull request:
Please see https://github.com/NixOS/nixpkgs/issues/148890 for an explanation of the problem.

I am submitting TWO PRs to close this issue; you may merge either one of them (it is not necessary to merge both)

This patch comes from https://gitlab.com/vicencb/kevinboot/-/blob/master/atf.patch

I have been running firmware with this patch on my daily driver rk3399
laptop for the past four months.

###### Motivation for this change

Please see https://github.com/NixOS/nixpkgs/issues/148890 for an explanation of the problem.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Commit:
remove unfree HDCP blob from arm-trusted-firmware, closes #148890

This patch comes from https://gitlab.com/vicencb/kevinboot/-/blob/master/atf.patch

I have been running firmware with this patch on my daily driver rk3399
laptop for the past four months.

lukegb commented:
For the record, the counterpart PR is #158309. Going to add to the discussion under #148890 rather than here, though, since it's more general about the philosophical change here.

a-m-joseph commented:
> the philosophical change here

It isn't a philosophical _change_; it corrects an oversight to bring licensing in line with longstanding policy.

I was very surprised to find CPU binaries without source code in a nixpkgs package whose license claimed `unfree=false`.  I think most people would be.  Otherwise what's the point?  The current licensing for this package is inconsistent with `pkgs/os-specific/linux/nvidia-x1/default.nix`.

a-m-joseph commented:
Note: I included the patch in the commit, rather than hash-referencing https://gitlab.com/vicencb/kevinboot/-/blob/master/atf.patch because minor changes to the ATF Makefile result in the original patch not applying correctly.

If you prefer I could find a way to minimize the committed patch so it only affects the Makfiles, and hash-reference vinceb's gitlab repo for the `.c` file changes.  But I figured if we have to have a patch in nixpkgs anyways we might as well put all the changes in one place.

The RK3399 firmware is extremely stable; Rockchip is focused on their newer chips at this point.  Any future merge conflicts should involve only the Makefiles (which ARM does seem to fiddle with now and then), and should be very straightforward to resolve.

a-m-joseph commented:
I will be updating this patch to take the approach recommended by @lukegb [here](https://github.com/NixOS/nixpkgs/issues/148890#issuecomment-1032002903) later tonight.

Commit:
arm-trusted-firmware: add unfreeIncludeHDCPBlob?true to control license and blob-removal patch

This change implements @lukegb's idea:

  https://github.gitop.top/NixOS/nixpkgs/issues/148890#issuecomment-1032002903

Specifically, it introduces a new parameter unfreeIncludeHDCPBlob
(defaults to true):

* If unfreeIncludeHDCPBlob==true then the license is changed to
  unfreeRedistributable, which will alert the user to the fact that
  the blob is being included (unless they set NIXPKGS_ALLOW_UNFREE=1).

* If unfreeIncludeHDCPBlob==false then the license is kept as bsd3, but
  a patch is applied to remove the HDCP blob from the build.

Commit:
make meta.license be a list rather than a single value

a-m-joseph commented:
On an amd64 host with:

`nix-build pkgsCross.aarch64-multiplatform.armTrustedFirmwareRK3399`

1. I have verified that nixpkgs will refuse to continue with NIXPKGS_ALLOW_UNFREE!=1

2. I am currently verifying a with `unfreeIncludeHDCPBlob=false`.  It is a large rebuild (I don't usually work off of `master`) which is why I am doing a crossbuild -- my amd64 machine has 32 cores and 64GB of memory and none of my arm64 machines are anywhere close to that.  It should be done by morning.

a-m-joseph commented:
Style feedback (even nitpicks) is welcome -- I have quite a number of patches that I'll be submitting over the next few weeks.

I managed to get mips64el to bootstrap (i.e. build nix out of nixpkgs) and have a bunch of other patches to make the systemd dependency of many packages optional on non-darwin kernels.

a-m-joseph commented:
Also let me know if you would prefer that I squash all three commits into a single commit.

a-m-joseph commented:
ping

lukegb commented:
Sorry for the delay, was hoping the listed maintainer would hop in with a review.

lopsided98 commented:
Sorry, I forgot about this PR. I wasn't aware of this blob and I think this is a good solution. cc @samueldr in case you're interested in this.

samueldr commented:
Shouldn't this affect only RK3399 builds?

a-m-joseph commented:
> I do not know if there's some magic property of this particular license that means it can't apply to binaries.

This [was discussed already](https://github.com/NixOS/nixpkgs/issues/148890#issuecomment-1031879737).  For software, the `license` field describes the license for the software's *source code*.

> Shouldn't this affect only RK3399 builds?

It only affects RK3399 builds.  No other platform uses that file.

Perhaps you were asking if the patch should be applied only when building for rk3399?  My understanding is that, in general, nixpkgs prefers that if a patch is to be applied and can be applied to all builds then it should be applied to all builds, so everything is building from the same source.

samueldr commented:
> It only affects RK3399 builds. No other platform uses that file.

No other platforms use that files, but the license change affect other platforms.

Currently this means that any non-RK3399 arm-trusted-firmware build is being marked as unfree by this change.

a-m-joseph commented:
> No other platforms use that files, but the license change affect other platforms.

That is why [I wrote](https://github.com/NixOS/nixpkgs/pull/158310#issuecomment-1138140403):

> One possibility would be to set the default value of `unfreeIncludeHDCPBlob` to `false` on all platforms except rk3399, since doing so on those platforms would not produce any change in the resulting output.

a-m-joseph commented:
https://github.com/NixOS/nixpkgs/pull/174686
